Sources/HummingbirdCompression/ResponseCompressionMiddleware.swift

```diff
@@ -0,0 +1,170 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Hummingbird server framework project
+//
+// Copyright (c) 2021-2024 the Hummingbird authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See hummingbird/CONTRIBUTORS.txt for the list of Hummingbird authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import CompressNIO
+import Hummingbird
+
+public struct ResponseCompressionMiddleware<Context: BaseRequestContext>: RouterMiddleware {
+    /// compression window size
+    let windowSize: Int
+    /// minimum size of response body to compress
+    let minimumResponseSizeToCompress: Int
+
+    /// Initialize ResponseCompressionMiddleware
+    /// - Parameters:
+    ///   - windowSize: Compression window size
+    ///   - minimumResponseSizeToCompress: Minimum size of response before applying compression
+    public init(
+        windowSize: Int = 65536,
+        minimumResponseSizeToCompress: Int = 1024
+    ) {
+        self.windowSize = windowSize
+        self.minimumResponseSizeToCompress = minimumResponseSizeToCompress
+    }
+
+    // ResponseBodyWriter that writes a compressed version of the response to a parent writer
+    class CompressedBodyWriter: ResponseBodyWriter {
+        let parentWriter: any ResponseBodyWriter
+        let context: Context
+        let compressor: NIOCompressor
+        var lastBuffer: ByteBuffer?
+
+        init(parent: any ResponseBodyWriter, context: Context, algorithm: CompressionAlgorithm, windowSize: Int) throws {
+            self.parentWriter = parent
+            self.context = context
+            self.compressor = algorithm.compressor
+            self.compressor.window = context.allocator.buffer(capacity: windowSize)
+            self.lastBuffer = nil
+            try self.compressor.startStream()
```

Comment:
Is this expected to never fail?

Reply:
There's no ! here. But it would only ever fail in the same situations the decompress would fail